### ISSUE TITLE
Jolting training metrics after running Model.evaluate() during Model.fit()

### ISSUE THREAD

**Issue by tom-andersson** [tensorflow/tensorflow#42994]
I'm copying some text over from [a Stack Overflow post of mine from last week](https://stackoverflow.com/questions/63711560/tf-keras-model-training-accuracy-jumps-after-running-model-evaluate-during-tra?noredirect=1#comment112664862_63711560) which is yet to be answered. I hope this is the right place to post for advice - I tried developers@tensorflow.org but my email was blocked.

Minimum working example demonstrating the problem on MNIST: https://colab.research.google.com/drive/1JLj23q5LjZu6cvcU7SQmfmY97qYwV5Gj?usp=sharing

I'm training a TensorFlow Keras `Model` using `Model.fit()`. I'm also using callbacks to log my training accuracy metrics after every batch using TensorFlow's `on_train_batch_end()` syntax. In addition, I'm using another callback to run `Model.evaluate()` every 500 batches to compute validation set accuracy and update the `logs` dict passed around the callbacks during `Model.fit()`.

Looking at the logged metrics vs. batch number shows very perplexing results. After the `Model.evaluate()` run, the training accuracy experiences a significant 'jolt', initially triggering a rapid increase in the logged training accuracy and subsequently triggering a significant drop training accuracy followed by a slower recovery. See the image below for an example.

![jolt-acc](https://user-images.githubusercontent.com/26459412/92324130-9bbd5c00-f036-11ea-9165-cb3e994dc46d.png)

My guess is that it's something to do with the Model.evaluate()'s call to `reset_metrics()`, which loops through and calls the reset_states() method on each metric. I can't work out what `reset_states()` is doing and if this is relevant to the behaviour I'm observing. Are the metrics shown during `Model.fit()` actually some form of moving averages rather than the batch-wise metric? In that case, the `reset_states()` method would be resetting the moving average, producing the jolting behaviour that flattens out over time.

I'm worried that calling `Model.evaluate()` within `Model.fit()` is silently training on my validation data, although I would expect better validation performance if that were the case.

**Comment by Saduf2019** [comment]
@tom-andersson 
Please provide complete stand alone code with tf version for us to replicate the issue faced or if possible share a colab gist with error reported.

**Comment by tom-andersson** [comment]
Hi @Saduf2019, I've included a Google Colab link after the first paragraph in my original post. This figure shows the same behaviour on MNIST. Printing the training and validation accuracies shows that they are not exactly equal to each other, but very close (equal to ~3 d.p.). Could this be the validation `logs` dict in `Model.evaluate()` corrupting the `logs` dict passed around during training perhaps?

![jolting_acc_colab](https://user-images.githubusercontent.com/26459412/92372153-6b3bf780-f0f4-11ea-8190-8ab6a7924420.png)

**Comment by tom-andersson** [comment]
Hi @Saduf2019 - I have now added the Google Colab link.

**Comment by Saduf2019** [comment]
@tom-andersson 
I do not find any colab link, please check and share it.

**Comment by tom-andersson** [comment]
@Saduf2019 
The link is after my first paragraph (I edited it in two mins before your post - I had forgotten initially, sorry).

Here it is: https://colab.research.google.com/drive/1JLj23q5LjZu6cvcU7SQmfmY97qYwV5Gj?usp=sharing

**Comment by Saduf2019** [comment]
I am able to replicate the issue reported,please find the [gist here](https://colab.research.google.com/gist/Saduf2019/e8b2e8bf3f2f4b717e46228fd2467f7a/untitled407.ipynb)

**Comment by omalleyt12** [comment]
@tom-andersson Thanks for the issue! You're right that calling `model.evaluate` during `model.fit` will reset the metrics. This won't affect training (your model won't train on the validation data), but it will cause the metrics to jump around the first few batches after calling `model.evaluate`. My suggestion would be to perform the `model.evaluate` call on epoch boundaries, since metrics are reset every epoch for `model.fit` anyway. You could always make each epoch smaller when using `tf.data` by passing `steps_per_epoch=500` to `model.fit`

**Comment by tom-andersson** [comment]
Thank you for clarifying @omalleyt12! Please can you point me to documentation where the metric averaging is expained? Is it a moving average of sorts? I couldn't find anything by Googling. What if I want the (noisy) batch-wise training metrics? Should I call `reset_metrics()` in `on_train_batch_begin()`?

Setting `steps_per_epoch` is not an ideal solution for me as `on_epoch_end()` will trigger the shuffling of samples in my data generator. The reason I want to validate every N batches is that one epoch is very slow with my dataset/architecture, and I want to model checkpoint within an epoch. I expect I'm not the only person who would appreciate clarity on the best way to do this with `tf.keras`.

Also, if my data generator didn't shuffle in `on_epoch_end()`, would the training never see the batches after `steps_per_epoch` batches?
